contracts/proposals/ccip020-graceful-protocol-shutdown.clar

```diff
@@ -0,0 +1,361 @@
+
+;; ERRORS
+
+(define-constant ERR_PANIC (err u2000))
+(define-constant ERR_VOTED_ALREADY (err u2001))
+(define-constant ERR_NOTHING_STACKED (err u2002))
+(define-constant ERR_USER_NOT_FOUND (err u2003))
+(define-constant ERR_PROPOSAL_NOT_ACTIVE (err u2004))
+(define-constant ERR_PROPOSAL_STILL_ACTIVE (err u2005))
+(define-constant ERR_NO_CITY_ID (err u2006)) ;; depracated in favor of MIA_ID and NYC_ID constants
+(define-constant ERR_VOTE_FAILED (err u2007))
+
+;; CONSTANTS
+
+(define-constant SELF (as-contract tx-sender))
+(define-constant MISSED_PAYOUT u1)
+(define-constant CCIP_020 {
+  name: "Graceful Protocol Shutdown",
+  link: "https://github.com/citycoins/governance/blob/feat/add-ccip-020/ccips/ccip-020/ccip-020-graceful-protocol-shutdown.md",
+  hash: "TBD",
+})
+
+(define-constant MIA_ID u1) ;; (contract-call? .ccd004-city-registry get-city-id "mia")
+(define-constant NYC_ID u2) ;; (contract-call? .ccd004-city-registry get-city-id "nyc")
+;; TODO: determine actual percentage from Miami
+(define-constant MIA_DRAW_PERCENTAGE u2) ;; 50% as placeholder
+
+;; MIA votes scaled to make 1 MIA = 1 NYC
+;; full calculation available in CCIP-020
+(define-constant VOTE_SCALE_FACTOR (pow u10 u16)) ;; 16 decimal places
+(define-constant MIA_SCALE_BASE (pow u10 u4)) ;; 4 decimal places
+(define-constant MIA_SCALE_FACTOR u8916) ;; 0.8916 or 89.16%
+
+;; DATA VARS
+
+;; vote block heights
+(define-data-var voteActive bool true)
+(define-data-var voteStart uint u0)
+(define-data-var voteEnd uint u0)
+
+;; start the vote when deployed
+(var-set voteStart block-height)
+
+;; DATA MAPS
+
+(define-map CityVotes
+  uint ;; city ID
+  { ;; vote
+    totalAmountYes: uint,
+    totalAmountNo: uint,
+    totalVotesYes: uint,
+    totalVotesNo: uint,
+  }
+)
+
+(define-map UserVotes
+  uint ;; user ID
+  { ;; vote
+    vote: bool,
+    mia: uint,
+    nyc: uint,
+  }
+)
+
+;; PUBLIC FUNCTIONS
+
+(define-public (execute (sender principal))
+  (let
+    (
+      (miaBalance (contract-call? .ccd002-treasury-mia-mining-v2 get-balance-stx))
+      (nycBalance (contract-call? .ccd002-treasury-nyc-mining-v2 get-balance-stx))
+      (miaRedemptionBalance (/ miaBalance MIA_DRAW_PERCENTAGE))
+      (nycRedemptionBalance nycBalance)
+    )
+
+    ;; check vote is complete/passed
+    (try! (is-executable))
+
+    ;; update vote variables
+    (var-set voteEnd block-height)
+    (var-set voteActive false)
+
+    ;; enable redemption extensions in the DAO
+    (try! (contract-call? .base-dao set-extensions
+      (list
+        {extension: .ccd012-redemption-mia, enabled: true}
+        {extension: .ccd012-redemption-nyc, enabled: true}
+      )
+    ))
+
+    ;; transfer funds to new redemption extensions
+    (try! (contract-call? .ccd002-treasury-mia-mining-v2 withdraw-stx miaRedemptionBalance .ccd012-redemption-mia))
+    (try! (contract-call? .ccd002-treasury-nyc-mining-v2 withdraw-stx nycRedemptionBalance .ccd012-redemption-nyc))
+
+
```

Comment:
For shutdown, asserts that the balances are zero after the transfers would be good.

Reply:
In general I agree doesn't hurt to have one more check. Could definitely do for NY but it depends on what happens with Miami, if they choose to pull anything out of the treasury it would mean the remaining amount is non-zero.

Reply:
Resolving this convo since the redemption code is removed from the CCIP.